cmd/deploy/input_test.go

```diff
@@ -2,9 +2,276 @@ package main
+					Host:       "localhost:5000",
+					Repository: "example",
+					Tag:        "bar",
+					Digest:     "sha256:25be82253336f0b8c4347bc4ecbbcdc85d0e0f118ccf8dc2e119c0a47a0a486e",
+				},
+			},
+			wantDigest: nil,
+			wantErr:    "",
+		},
+		{
+			name: "basic fallback",
+			rawLookup: map[string]string{
+				"": "example",
+			},
+			wantLookup: map[ociregistry.Digest]registry.Reference{
+				"": {
+					Host:       "docker.io",
+					Repository: "library/example",
+				},
+			},
+			wantDigest: nil,
+			wantErr:    "",
+		},
+		{
+			name: "Digest fallback",
+			rawLookup: map[string]string{
+				"": "example@sha256:25be82253336f0b8c4347bc4ecbbcdc85d0e0f118ccf8dc2e119c0a47a0a486e",
+			},
+			wantLookup: map[ociregistry.Digest]registry.Reference{
+				"": {
+					Host:       "docker.io",
+					Repository: "library/example",
+					Digest:     "sha256:25be82253336f0b8c4347bc4ecbbcdc85d0e0f118ccf8dc2e119c0a47a0a486e",
+				},
+			},
+			wantDigest: getDigest(""),
+			wantErr:    "",
+		},
+		{
+			name: "Tag fallback",
+			rawLookup: map[string]string{
+				"": "example:tag",
+			},
+			wantLookup: map[ociregistry.Digest]registry.Reference{
+				"": {
+					Host:       "docker.io",
+					Repository: "library/example",
+					Tag:        "tag",
+				},
+			},
+			wantDigest: getDigest(""),
+			wantErr:    "",
+		},
+		{
+			name: "bad digest",
+			rawLookup: map[string]string{
+				"invalid digest": "example:foo",
+			},
+			wantErr: `lookup key "invalid digest" invalid:`,
+		},
+		{
+			name: "bad reference",
+			rawLookup: map[string]string{
+				"sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d": "://#",
+			},
+			wantErr: "failed to parse lookup ref",
+		},
+		{
+			name: "digest mistmatch",
+			rawLookup: map[string]string{
+				"sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d": "example:foo@sha256:00f42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d",
+			},
+			wantErr: "digest on lookup ref should either be omitted or match key:",
+		},
+	} {
+		t.Run(x.name, func(t *testing.T) {
+			lookups, lookupDigest, err := normalizeInputLookup(x.rawLookup)
+
+			if x.wantErr != "" {
+				if err == nil {
+					t.Fatalf("Expected error not returned: %s", x.wantErr)
+				}
+
+				if !strings.Contains(err.Error(), x.wantErr) {
+					t.Fatalf("Expected error doesn't match.\ngot:\n%q,\n\nexpected to contain:\n%q", err, x.wantErr)
+				}
+			}
+
+			if x.wantDigest != nil && lookupDigest == nil {
+				t.Errorf("Unexpected last digest.\ngot:\n%s\n\nexpected:\n%s\n", "<nil>", x.wantDigest)
+			} else if x.wantDigest == nil && lookupDigest != nil {
+				t.Errorf("Unexpected last digest.\ngot:\n%s\n\nexpected:\n%s\n", lookupDigest, "<nil>")
+			} else if lookupDigest != nil && x.wantDigest != nil && *lookupDigest != *x.wantDigest {
+				t.Errorf("Unexpected last digest.\ngot:\n%s\n\nexpected:\n%s\n", lookupDigest, x.wantDigest)
+			}
```

Comment:
Since all these cases are explicit anyhow, I think we should go back to what I suggested in https://github.com/docker-library/meta-scripts/pull/39#discussion_r1583831223, except maybe reverse it (explicit `"nil"` instead of `"empty"`) so the string comparison is simpler and we can avoid the pointers (and the `getDigest` function).  Also, I think we can collapse them (and reverse the conditionals to match the message order so there's less cognitive load while reading it):

```suggestion
			if (lookupDigest == nil && x.wantDigest != "nil") || (lookupDigest != nil && x.wantDigest == "nil") || (lookupDigest != nil && *lookupDigest != x.wantDigest) {
				t.Errorf("Unexpected digest.\ngot:\n%v\n\nexpected:\n%v\n", lookupDigest, x.wantDigest)
			}
```

Reply:
The `%v` was the part I was missing. It was expanded because I couldn't get `<nil>` to display properly.